### DISCUSSION TITLE
Plot variable over line vs. PointValue PostProcessor

### DISCUSSION THREAD

**Discussion by heinono1** [idaholab/moose#23679] [Q&A General] [answered]
Hi. I am running micromagnetic simulations using the Ferret modules. As I work in industry designing magnetic writers for magnetic disk drives, it is imperative that I collect accurate values of quantities such as magnetostatic fields generated from volume and surface magnetic charges. As part of the simulations, I solve the Poisson equation for the magnetostatic potential, and use an AuxKernel (really just computing the negative gradient of the potential) to obtain the magnetostatic field from the potential. The simulations of course dump the magnetostatic potential and the magnetostatic field in the Nemesis files, but I also use a PointValue PostProcessor to dump the magnetostatic field at some particularly important points. I noticed that the field values that are output by the PostProcessor are larger by 3%-4% than an old code (Magpar) that we are thinking of retiring. That was a little odd since John Mangeri and I have gone through great pains to ensure that Ferret calculates the magnetostatic field correctly. But then I compared the values of the Aux variables for the magnetostatic field as output in the Nemesis files with those output by the PointValue postprocessor, and they are not the same: the PointValue postprocessor values are larger in magnitude by 3%-4%! This is odd. Is there some complicated reason why this might occur? The nonlinear variable on which the Aux kernel acts is a LAGRANGE variable so the Aux variables are nodal.

Thanks,
Olle

**Comment by GiudGiud** [comment, marked as answer]
Hello

PointValue relies on this routine
https://libmesh.github.io/doxygen/classlibMesh_1_1System.html#afbcbcceb6ee4f02272c69d5292394603
to get the value, which relies on the one below that evaluates the shape functions. 

Are you using displaced meshes?
@roystgnr I m not sure why they would differ but you have more experience in how our exodus output works

Guillaume

**Reply by heinono1** [reply]
Thanks, Guillaume. I am not using displaced mesh. Interesting to see that the PointValue is expensive (the search algorithm for the element). Maybe better just to use Paraview or other postprocessing of the Nemesis/Exodus files by creating a small mesh block for the region of interest?

Cheers,
Olle

**Reply by GiudGiud** [reply]
It s expensive mostly the first time, when the point locator gets created if it has not been already

**Reply by heinono1** [reply]
Thanks - that makes sense. But I am still puzzling over the difference between the PointValue and the Exodus Plot over line value....

**Reply by heinono1** [reply]
I am going to mark this answered (thanks, Guillaume). I believe the discrepancy just has to do with the fact that the output in the Exodus file is a bit noisy (nodal values and and finite mesh resolution) and how the postprocessor evaluates the variables at the specified points - it seems really just to boil down to spatial noise because of the finite mesh resolution. I believe I can filter the results using custom postprocessing outside of MOOSE.

**Reply by GiudGiud** [reply]
Roy is on vacation and he knows a lot more than me btw

**Reply by roystgnr** [reply]
The point locator creation is something like O(N log N), but even point locator use is O(log N) with IIRC an ugly constant, not something you'd want to use unnecessarily.  It's as good as you can get asymptotically if you don't already know what element you're on, but it's bad enough that we have a bunch of code which looks like "try a good guess about what element you're on, then fall back on the point locator if we guessed wrong".

If you're using LAGRANGE, that rules out my first hypothesis for the discrepancy, about how we have to decimate higher-order solutions to fit onto the second-order elements in Exodus.  Exodus output of linear and quadratic Lagrange elements is pretty much one-to-one.  Although ... in my experience Paraview will further decimate second-order to piecewise-first-order when displaying a solution; is it possible they do the same when postprocessing a solution as well??

Maybe you're hitting a loose tolerance on the point locator if you're near an element boundary, getting an evaluation from a neighboring element?  But the default relative tolerance is like 1e-6, and that times your flux jump (plus some even-higher-order terms) shouldn't be anywhere near 3e-2.

I'd love to see an example of the failure.  If there's a MOOSE/libMesh bug then we ought to be fixing it; if there's a Paraview issue we ought to be letting users know about it.  There's a level at which I too throw up my hands up and say "floating point, noise, whatcha gonna do", but it's at more like 1e-10, not 3e-2.

**Reply by heinono1** [reply]
Thanks, Roy. I am using LAGRANGE. I dug in a little bit more and compared the MOOSE postprocessor output (PointValue) with Paraview's rendering from the Exodus file, and also using Paraview to calculate the gradient of the potential in the Exodus file. It turns out that the potential in Exodus that Paraview renders is very smooth, and the Paraview-calculated gradient is too - smoother than the Moose postprocessor output, for some reason. The Paraview rendering of the Exodus field is rather choppy. It may also be that the particular point at which I want the field has some pathology associated with it (like being at an element boundary) - at that point, the PointValue postprocessor field is always some 3%-4% larger than it "should" be (based, for example, on the Paraview-calculated gradient of the potential). I should probably re-mesh that particular block to see if changes the output field values at that particular point.
